keps/sig-node/1592-pod-level-resource-limits/README.md

```diff
@@ -0,0 +1,439 @@
+
+### Goals
+
+This proposal aims to:
+
+* Allow developers to define resource limits on the pod level in addition to the individual container level.
+* Use as much as possible the Linux kernel's cgroup ability to define limits in hierarchies to enforce the defined limits.
+
+### Non-Goals
+
+*  Providing a general-purpose interface to the full range of possible resource management provided by the Linux cgroup hierarchy.
+
+## Proposal
+
+The Pod QoS enhancement already implemented in Kubernetes manages resources as a hierarchy of cgroups in the following way:
+
+<pre>
+  kubepods
+   |
+   +-- Guaranteed-pod0
+   |   |
+   |   +-- container0 (pause container)
+   |   |
+   |   +-- container1 (first container)
+   |   |
+   |  = ... =
+   |   |
+   |   +-- containerN (N-th container)
+   |
+   +-- QoS CGroup (one of burstable, or besteffort)
+       |
+       \ pod 
+          |
+          +-- container0 (pause container)
+          |
+          +-- container1 (first container)
+          |
+        = ... =
+          |
+          +-- containerN (N-th container)
+</pre>
+
+The proposal in this KEP is to allow users to define resources on the Pod level itself. Both resource requests and limits are allowed. The scheduler
+algorithm will consider the pod-level resources if they exist in preference to the container-level resources. If no pod-level resource section is provided,
+then the scheduler will continue to operate as it does in the current implementation. The QoS definitions will be updated to consider the pod-level resources
+when classifying a pod into one of the QoS levels. If the pod level doesn't include any resource limits, then kubelet will continue to function as it does today.
+
+The current implementation will set the cgroup limits for the memory/CPU resources on the pod level of the hierarchy only in the following cases:
+1. The QoS level is `Guaranteed`. 
+1. The QoS level is `Burstable` and all of the containers specify a limit for the relevant resource. 
+
+In the current implementation pods are considered `Guaranteed` if and only if all of the included containers define a resource request which is equal
+to the resource limit. The current implementation considers a pod to be `Burstable` if at least one container defines resource requests or limits. Any
+pods not matching the `Guaranteed` or `Burstable` QoS definitions are placed in the `BestEffort` QoS level.
+
+This proposal suggests modifying these definitions the following: 
+
+A pod is considered `Guaranteed` if either of the following two conditions apply: 
+1. All of the containers included in the pod define memory and CPU resource requests equal to resource limits
+1. The pod includes a pod-level memory and CPU resource request which is equal to the resource limit, even if one or more of the included containers doesn't define either a CPU or a memory request or limit.
```

Comment:
Should we address how will pod level resource limits affect [ephemeral containers](https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#ephemeral-container)?